### DISCUSSION TITLE
Support for other languages

### DISCUSSION THREAD

**Discussion by RandomBrown** [SeriousOldMan/Simulator-Controller#12] [Q&A] [answered]
Hi. How can I install the Spanish language (es-ES) for both voice recognition and TTS?
My Windows 10 is in Spanish, and the speech recognition works fine in other programs, but the Simulator-Controller does not recognize it. Only German and English appear. No support for other languages?

I have followed the steps indicated in "Tab Voice Control", and the installed package MSSpeech es-ES can be indicated for voice recognition (I don't know why it doesn't recognize the language pack that is already installed with Windows 10 and that the rest of the programs). But under "Language" within "Voice Control", only the German and English languages still appear.

Could you make a short guide indicating the steps necessary for both voice recognition and TTS to be in another language? Or use the language packs that are already configured in Windows 10.

Thanks!

**Comment by SeriousOldMan** [comment, marked as answer]
Hi, great question.

There are language packs for the Windows voice recognition available, both for the local engine as well as for the Azure cloud based version. But - you also have to have the recognition grammatics and spoken phrases translated to the target language. And I only managed to have an english and german version up to now, understandable, since these are the languages, I am comfortable with. You can find the grammar files here:

 https://github.com/SeriousOldMan/Simulator-Controller/tree/main/Sources/Assistants/Grammars
 
 Maybe you want to become a contributor and manage the translation for the spanish language?
 
 Best regards

Oliver

**Comment by RandomBrown** [comment]
Oh! I accept the challenge! ;) I'm going to start with those files. Give me just a little time.
Congratulations on such a great tool!

**Comment by RandomBrown** [comment]
In this example:
PitstopNoTyreChange=(CanWe) {leave, keep} the tyres on the car

Should I translate the items in parentheses and braces as well, or are those items substituted and translated elsewhere and not in that sentence?

**Reply by SeriousOldMan** [reply]
(CanWe) is a fragment, a kind of variable. You will find it in the upper part of the file.

{leave, keep} are alteratives in a given sentence in the grammar, whereas

[sentence1... , sentence2] are completey separate sentences with the same meaning.

Example:

[ {Can you, Please} give me the screwdriver, I need the screwdriver]

This has 3 valid inputs:

1. Can you give me the screwdriver
2. Please give me the screwdriver
3. I need the screwdriver

Hope, this helps...

If you manage to make a valid translation, I will add it to the distribution. But I may need your help for future updates. :-)

Cheers, Oliver

**Reply by RandomBrown** [reply]
Clear! Count on my ax!

It is a pleasure to collaborate on your project. It is magnificent.

But then, would this be correct?
Original (en)--> PitstopNoTyreChange=(CanWe) {leave, keep} the tyres on the car
Translate (es)--> PitstopNoTyreChange=(Podemos) {dejar, mantener} estos neumáticos

Or should it be?
Translate--> PitstopNoTyreChange=(CanWe) {dejar, mantener} estos neumáticos

**Reply by SeriousOldMan** [reply]
Up to you. "CanWe" is only a label. I think it will help you, when you compare the english grammer file with the german one....

**Comment by SeriousOldMan** [comment]
Further explanation:

CanWe is a substitute label, a kind of macro. Substitute labels are defined in the "[Choices]" section. You can define as many as you like and also you can name it as you like.

But you must leave all left side names in the "[Fragments]" section in english, since they are kind of adressed from the programm code. But you have to translate the right side. "Fragments" are used in the output phrases of the assistants, for example:

PressureCorrectionUp.2=Since the temperatures are falling we have increased the pressures by %value% %unit% for the next stint.

**Reply by RandomBrown** [reply]
But every time I find a substitute label, for example "(CanWe)", do I have to translate it, or just do it when it appears in the "[Choices]" section?

**Comment by SeriousOldMan** [comment]
Only in the Choices section. You can name the label "CanWe" in spanish as well. It is only the label for the alternative phrases in the [Choices] section. But you can also translate "CanWe" to "Podemos". Might be clearer for the reader.

**Comment by SeriousOldMan** [comment]
Forgot to mention. Maybe you want to join our Discord, unless you've already done that.

**Comment by SeriousOldMan** [comment]
Hi, how is it going. Do you already talk in Spanish to the Race Assistants?

**Comment by RandomBrown** [comment]
Hello! I'm sorry for the delay. After the Christmas holidays, I get back to the routine. I already have both files translated, and I am testing them. At the moment I can't get a good voice recognition so that I can quickly test the different combinations of questions and answers.

Of course, I will join your Discord!

**Comment by SeriousOldMan** [comment]
Hi, recognition quality depends heavily on the "correct" microphone level. It also helps to use the training function of Windows, so that Windows knows your articulation.

Ciao, Oliver

**Comment by SeriousOldMan** [comment]
Hi, I want to ask again, how is it going?  I've shared the fact, that someone is working on a spanish translation on our Discord and there is quite some interest! You can make you a name as a contributor :-)

**Comment by RandomBrown** [comment]
Sorry for the delay. I have little free time and it is difficult to combine everything.
I am finishing the test. I have changed some expressions and words to make them more understandable in Spanish.
Upgrading to v3.8.4 freezes when "updating translations to Release 3.8.4".
I'm going to uninstall and do a clean install. I'll keep you informed.

**Comment by SeriousOldMan** [comment]
Hi, the release yesterday introduced a third assistant, a spotter. You will find the grammar files for this assistant at the usual location. But I recommend to wait with a translation one or two more releases, since there will be a lot phrases added in the coming releases.

By the way: Have you fixed your "Freeze". If not, I am here to help...

**Comment by SeriousOldMan** [comment]
Hi, long time no see :-)

I just want to inform you, that an initial version of a Spanish translation will be released by the end of this week thanks to the great effort of another user.

**Comment by SeriousOldMan** [comment]
Just stumbled upon our conversatio here and want to let you know, that a Spanish translation has been made available in the meantime. Maybe of interest for you.
